vignettes/custom_epiworkflows.Rmd

```diff
@@ -0,0 +1,554 @@
+
+## Define the `frosting()`
+Since the post-processing frosting layers[^1] are unique to this package, to
+inspect them we use `extract_frosting()` from `{epipredict}`:
+
+```{r inspect_fwa_layers, warning=FALSE}
+epipredict::extract_frosting(four_week_ahead$epi_workflow)
+```
+
+The above gives us detailed descriptions of the arguments to the functions named
+above in the postprocessor of `four_week_ahead$epiworkflow`.
+Creating the layers is a similar process, except with frosting instead of a
+`recipe`[^2]:
+
+```{r make_frosting}
+four_week_layers <- frosting() |>
+  layer_predict() |>
+  layer_residual_quantiles(quantile_levels = c(0.1, 0.25, 0.5, 0.75, 0.9)) |>
+  layer_add_forecast_date() |>
+  layer_add_target_date() |>
+  layer_threshold()
+```
+
+
+Most layers will work for any engine or steps; `layer_predict()` you will want
+to call in every case.
+There are a couple of layers, however, which depend on whether the engine used predicts quantiles or point estimates.
+
+The layers that are only supported by point estimate engines (such as
+`linear_reg()`) are
+
+- `layer_residual_quantiles()`: the preferred method of generating quantiles for
+  non-quantile models, it uses the error residuals of the engine. 
+  This will work for most parsnip engines.
+- `layer_predictive_distn()`: alternate method of generating quantiles, it uses
+  an approximate parametric distribution. This will work for linear regression
+  specifically.
+  
+TODO check this
+  
+On the other hand, the layers that are only supported by quantile estimating
+engines (such as `quantile_reg()`) are
+
+- `layer_quantile_distn()`: adds the specified quantiles.
+  If they differ from the ones actually fit, they will be interpolated and/or
+  extrapolated.
+- `layer_point_from_distn()`: this adds the median quantile as a point estimate,
+  and should be called after `layer_quantile_distn()` if called at all. 
+
+## Fitting an `epi_workflow()`
+
+Given that we now have a recipe and some layers, we need to assemble the workflow:
+```{r workflow_building}
+four_week_workflow <- epi_workflow(
+  four_week_recipe,
+  linear_reg(),
+  four_week_layers
+)
+```
+
+And fit it to recreate `four_week_ahead$epi_workflow`
+```{r workflow_fitting}
+fit_workflow <- four_week_workflow |> fit(training_data)
+```
+
+## Predicting
+
+To do a prediction, we need to first narrow the dataset down to the relevant
+datapoints:
+
+```{r grab_data}
+relevant_data <- get_test_data(
+  four_week_recipe,
+  training_data
+)
+```
+
+With a fit workflow and test data in hand, we can actually make our predictions:
+
+```{r workflow_pred}
+fit_workflow |> predict(relevant_data)
+```
+
+Note that if we had simply plugged `training_data` into `predict()` we still get
+predictions:
+
+```{r workflow_pred_training}
+fit_workflow |> predict(training_data)
+```
+
+The resulting tibble is 800 rows long, however.
+This produces forecasts for not just the actual `forecast_date`, but for every
```

Comment:
question: where did we set the `forecast_date`/how does the workflow know what it is? what if we want to use a different forecast date? do we have to re-define and re-compile the whole workflow?

Reply:
This is set in `epi_recipe`. I'll make some changes to account for this. Also I'll have to change the name to `reference_date`, since that's what's in that function.